test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,185 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/iancoleman/strcase"
+	ispnv1 "github.com/infinispan/infinispan-operator/api/v1"
+	"github.com/infinispan/infinispan-operator/api/v2alpha1"
+	"github.com/infinispan/infinispan-operator/controllers"
+	"github.com/infinispan/infinispan-operator/controllers/constants"
+	batchtest "github.com/infinispan/infinispan-operator/test/e2e/batch"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	"github.com/operator-framework/api/pkg/manifests"
+	coreos "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	ctx                   = context.TODO()
+	testKube              = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	subName               = constants.GetEnvWithDefault("SUBSCRIPTION_NAME", "infinispan-operator")
+	subNamespace          = constants.GetEnvWithDefault("SUBSCRIPTION_NAMESPACE", tutils.Namespace)
+	catalogSource         = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE", "test-catalog")
+	catalogSourcNamespace = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE_NAMESPACE", tutils.Namespace)
+	subPackage            = constants.GetEnvWithDefault("SUBSCRIPTION_PACKAGE", "infinispan")
+
+	packageManifest = testKube.PackageManifest(subPackage, catalogSourcNamespace)
+	sourceChannel   = getChannel("SUBSCRIPTION_CHANNEL_SOURCE", 1, packageManifest)
+	targetChannel   = getChannel("SUBSCRIPTION_CHANNEL_TARGET", 0, packageManifest)
+
+	subStartingCsv = constants.GetEnvWithDefault("SUBSCRIPTION_STARTING_CSV", sourceChannel.CurrentCSVName)
+)
+
+func TestUpgrade(t *testing.T) {
+	printManifest()
+	testKube.NewNamespace(tutils.Namespace)
+
+	replicas := 2
+	sub := &coreos.Subscription{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: coreos.SubscriptionCRDAPIVersion,
+			Kind:       coreos.SubscriptionKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      subName,
+			Namespace: subNamespace,
+		},
+		Spec: &coreos.SubscriptionSpec{
+			Channel:                sourceChannel.Name,
+			CatalogSource:          catalogSource,
+			CatalogSourceNamespace: catalogSourcNamespace,
+			InstallPlanApproval:    coreos.ApprovalManual,
+			Package:                subPackage,
+			StartingCSV:            subStartingCsv,
+		},
+	}
+	testKube.CreateSubscription(sub)
```

Comment:
`OperatorGroup` should be also created unless it's global namespace. If the OperatorGroup doesn't exist the installation won't proceed on OpenShift. It's interesting that it passes on Kind in CI though

Reply:
The `OperatorGroup` is actually created as part of `scripts/ci/install-catalog-source.sh`. I did this originally as it helped with manual testing as well, but I can move the creation to the `upgrade-test.go` if you prefer?

Reply:
Maybe it would be better so the test is more self-contained? I mean the test creates and removes the namespace where it installs the Operator by default. If the user will only rerun the test it will fail on second attempt as the OperatorGroup will be removed.